pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/internal/LatticeRelation.java

```diff
@@ -0,0 +1,339 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.rule.internal;
+
+import static java.util.stream.Collectors.toSet;
+import static net.sourceforge.pmd.util.CollectionUtil.any;
+import static net.sourceforge.pmd.util.CollectionUtil.finish;
+import static net.sourceforge.pmd.util.CollectionUtil.map;
+import static net.sourceforge.pmd.util.CollectionUtil.toMutableList;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collector;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.internal.util.AssertionUtil;
+import net.sourceforge.pmd.lang.rule.internal.GraphUtils.DotColor;
+
+/**
+ * Indexes data of type {@code <V>} with keys of type {@code <K>}, where
```

Comment:
Just a question: Do we really need to be that generic? We are dealing with AST Nodes, don't we?
That's btw one example for #2611 - it will be very difficult to understand this class without looking at many more classes in parallel to understand, what's going on....
Maybe the question in a different way: Do we really implement it that abstract? We only need it for PMD purposes, not for general purpose....

Reply:
Ok, as far as I understand, that is your solution for dealing with Abstract node types.... right?

Reply:
> Ok, as far as I understand, that is your solution for dealing with Abstract node types.... right?

Yes that's it.

Honestly making it generic is easy and makes tests much more comprehensive and easier to write. This data structure is a kind of multimap where values are added at once to several keys, doing minimal work. So we can add `Node` values to several `Class` keys at once, just by knowing `node.getClass()`, without having to traverse all the supertypes for every instance. We also don't accumulate values for keys that won't be queried (those `Class` instances for which no rulechain rule has added a visit request).


The API is really simple I think, you just use `put` and `get` as if it was a regular map.